### PR TITLE
Generate message lookup in gen_l10n

### PR DESCRIPTION
**Pull request by HansMuller**
The gen_l10n tool now generates code that looks up locale-specific message translations, rather than depending on the Dart intl package. Doing so will make it possible for modules (typically packages) to bundle their own localizations, and it will speed up build times for internationalized apps.

Removed a few gen_l10n tests that checked the text generated code for equality. Those tests have already been made obsolete by integration tests from https://github.com/flutter/flutter/pull/49586.

In addition to the functionality changes, there is quite a bit of refactoring and some gratuitous renaming.

The approach taken here is similar to how Flutter's flutter_localizations package manages translations for the Flutter SDK itself.

### CODE REVIEW

File: dev/tools/localization/gen_l10n.dart [hunk 1]
```diff
@@ -295,67 +109,190 @@ String generatePluralMethod(Message message) {
...
   /// Creates an instance of the localizations generator class.
   ///
   /// It takes in a [FileSystem] representation that the class will act upon.
   LocalizationsGenerator(this._fs);
 
-  static RegExp arbFilenameLocaleRE = RegExp(r'^[^_]*_(\w+)\.arb$');
-  static RegExp arbFilenameRE = RegExp(r'(\w+)\.arb$');
-
   final file.FileSystem _fs;
+  // TBD: private data
```
Comment (shihaohong) [inline dev/tools/localization/gen_l10n.dart]: Why is this TBD? Does it need to be public for some reason?

File: dev/tools/localization/gen_l10n.dart [hunk 2]
```diff
@@ -295,67 +109,190 @@ String generatePluralMethod(Message message) {
...
+          .replaceAll('@(languageCode)', localesForLanguage.languageCode)
+          .replaceAll('@(class)', '$className${LocaleInfo.fromString(localesForLanguage.languageCode).camelCase()}')
+          .replaceAll('@(switchClauses)', locales.map((LocaleInfo locale) {
+            return switchClauseTemplate
+              .replaceAll('@(case)', countryCode)
+              .replaceAll('@(class)', '$className${locale.camelCase()}');
+          }).join('\n'));
+      }).join('\n');
+    }
+    assert(false, 'TBD dispatch on scriptCodes, scriptCodes AND countryCodes');
```
Comment (shihaohong) [inline dev/tools/localization/gen_l10n.dart]: Is it ok to leave this as an assert? It we do, maybe we should rephrase it to suggest what is possible.

Also, when should we expect the tool to handle script codes as well?

**Review by shihaohong — COMMENTED** [review]
~Only partway done (3/5 files)~, a few nits here and there and questions regarding TBD stuff, but overall, this looks really good ~so far~!

Edit: Fully done, turns out the remaining 2 files were simple to look over

File: dev/tools/localization/gen_l10n_templates.dart [hunk 3]
```diff
@@ -152,21 +93,103 @@ class @(className) {
...
+@(numberFormatting)
+    return @(message);
+  }''';
+
+const String pluralMethodTemplate = '''
+  @override
+  String @(name)(@(parameters)) {
+@(dateFormatting)
+@(numberFormatting)
+    return intl.Intl.pluralLogic(
```
[the remark this answers is not in the document]
Comment (HansMuller) [reply dev/tools/localization/gen_l10n_templates.dart]: The `Intl.plural` method is like `Intl.message`, it relies on  intl_translation:extract_to_arb. This PR is about eliminating that dependency.

File: dev/tools/localization/gen_l10n.dart (hunk 1 above)
Comment (HansMuller) [reply dev/tools/localization/gen_l10n.dart]: Oops, I'd meant to make these fields private.

File: dev/tools/localization/gen_l10n.dart (hunk 2 above)
Comment (HansMuller) [reply dev/tools/localization/gen_l10n.dart]: It should handle script codes, although we don't have tests cases to cover that yet. I will add this stuff; didn't mean to leave it as a TBD.

**Review by shihaohong — APPROVED** [review]
LGTM
